Lesson 5

imh0t3mp opened the pull request:
Пффф, готово.
1. Убрал абстрактный класс. Лишняя сущность, лишние заморочки. Достаточно простого базового класса Fruit
2. Переименовал некоторые методы для оперирования содержимым ящиков - putOneItem -> putItem
3. Перемещение содержимого из одного ящика в другой не имеет смысла в классе Box. Задачу перемещения должен реализовывать сервис оперирования объектом а не объект. Но сам метод оставил и в классе Box, хоть он и нарушает, по моему мнению, S принцип SOLID  
4. StockService -> оперирует Box и Stock
Stock -> оперирует только своим списком, элементы хранимые в сиске не меняет 
Box -> оперирует только своим списком, элементы списка не меняет
Fruit (Apple,Orange...) - полностью закрыты для изменения.
5. Ящик может вернуть свой вес
6. Ящик может сравнить свой вес с весом другого ящика
7. Добавил набор типических тестов которые, по чести говря, не охватывают всего что было заложено.
8. Хоть всё и сделано на List, для приведения к действительности, добавил граничения на объём ящика.
9. Никаких Queue, только List

Commit:
Первый этап решения домашней работы.
Добавил вариант хранения данных не в массиве а в коллекции ArrayList
Для обратной совместимости с хранилищем на массиве, метод getAllTasks кастует List в массив Task[]
На типовом наборе тестов всё работает

Commit:
Рефакторинг. Вынес всё связанное с трекером задач в отдельный пакет

Commit:
Замечательно, есть Яблоки, Апельсины и просто фрукты.
Есть ящики, которые могут хранить только фрукты как таковые или что-то из фруктов.
На всякий случай, добавил к каждому фрукту унимкальный UUID, ну так, чтоб отличать одно яблоко от другого.

Commit:
Перенёс генерацию генома фрукта в абстрактный класс.
Для ящика добавил генерацию уникального ID ящика и набор методов для сравнения ящиков между собой.

Добавил минимальный набор исключений, которорые могут, по идее, возникнуть при оперировании с ящиками и складом
Теперь вместо StockService есть класс описывающий склад на ArrayList. Класс реализует интерфейс Stock, где описаны все необходимые операции, которые реализует склад
Все операции со складом будут реализованы в рамках класса StockService

Commit:
Ммм, сделано многое.
ДОбавил сравнения ящиков по весу, добавил операции с ящиками на складе через сервис. И прочее

Commit:
TODO: надо разобраться, что я там понаворотил со сравнением и абстрактными классами

Commit:
Ммм, надеюсь работает

Commit:
Пффф, готово.
1. Убрал абстрактный класс. Лишняя сущность, лишние заморочки. Достаточно простого базового класса Fruit
2. Переименовал некоторые методы для оперирования содержимым ящиков - putOneItem -> putItem
3. Перемещение содержимого из одного ящика в другой не имеет смысла в классе Box. Задачу перемещения должен реализовывать сервис оперирования объектом а не объект. Но сам метод оставил и в классе Box, хоть он и нарушает, по моему мнению, S принцип SOLID
4. StockService -> оперирует Box и Stock
Stock -> оперирует только своим списком, элементы хранимые в сиске не меняет
Box -> оперирует только своим списком, элементы списка не меняет
Fruit (Apple,Orange...) - полностью закрыты для изменения.
5. Ящик может вернуть свой вес
6. Ящик может сравнить свой вес с весом другого ящика
7. Добавил набор типических тестов которые, по чести говря, не охватывают всего что было заложено.
8. Хоть всё и сделано на List, для приведения к действительности, добавил граничения на объём ящика.
9. Никаких Queue, только List

Commit:
Не удержался, удалил метод чтоб он глаза не мозолил. Не место этому методу тут.

JavaStoryTeller commented on lesson_5/src/main/java/name/imh0t3mp/course/geekbrains/fruit_stock/Box.java:
```diff
@@ -0,0 +1,170 @@
+     * @return - один фрукт
+     * @throws NoItemInBox - если в ячейке нет фрукта
+     */
+    public T getItem(int index) throws NoItemInBox {
```
Тут спокойно можно обойтись без перехвата исключения, обычными проверками.
if (index >= fruitsInBox.size()) {
  throw new NoItem...
}
return ...

imh0t3mp replied:
Согласен

JavaStoryTeller commented on lesson_5/src/main/java/name/imh0t3mp/course/geekbrains/fruit_stock/Apple.java:
```diff
@@ -0,0 +1,9 @@
+
+public class Apple extends Fruit {
+    public Apple() {
+        this.weight = 1.0f;
```
Если commonType всегда равен имени класса, то чтобы не дублировать это в конструкторах наследников, можно вынести эту строку в конструктор родителя

imh0t3mp replied:
Согласен. В данном случае это упростит код.

JavaStoryTeller commented on lesson_5/src/main/java/name/imh0t3mp/course/geekbrains/fruit_stock/Box.java:
```diff
@@ -0,0 +1,170 @@
+     * @param box - ящик для сравнения
+     * @return - TRUE|FALSE Одинаковы ли ящики по весу
+     */
+    public boolean compareTo(Box box) {
```
Некорректное сравнение float'ов

imh0t3mp replied:
Учёл это замечание. Эта и несколько других правок в новом коммите.

Commit:
Учёл замечания:
1 - commonType определяю не по имени класса а явно задаю в конструкторе;
2 - при сравнении веса ящиков учитываю точность;
3 - при получении элемента из коробки, сразу проверяю ёмкость коробки, вместо обработки сключения.

Немного расширил набор проверок

imh0t3mp commented:
Задачу оценили. Было несколько замечаний, которые я постарался учесть.

imh0t3mp committed:
Merge branch 'master' into lesson-5
